src/charmed_kubeflow_chisme/components/pebble_component.py

```diff
@@ -16,28 +15,128 @@
 logger = logging.getLogger(__name__)
 
 
-@dataclass
-class ContainerFileTemplate:
-    """Dataclass for defining templates that should be rendered as files in Pebble Containers."""
+class LazyContainerFileTemplate:
+    """A lazy container file template that can be rendered and pushed into a Pebble container."""
 
-    source_template_path: Union[Path, str]
-    destination_path: Union[Path, str]
-    context_function: Optional[Union[Callable, dict]] = None
-    user: Optional[str] = None
-    group: Optional[str] = None
-    permissions: Optional[str] = None
+    def __init__(
+        self,
+        destination_path: Union[Path, str],
+        source_template_path: Optional[Union[Path, str, Callable[[], Union[Path, str]]]] = None,
+        source_template: Optional[Union[str, Callable[[], str]]] = None,
+        context: Optional[Union[dict, Callable[[], dict]]] = None,
+        user: Optional[str] = None,
+        group: Optional[str] = None,
+        permissions: Optional[str] = None,
+    ):
+        """A ContainerFileTemplate that lazily generates its template from a path or string."""
+        if source_template_path is not None and source_template is not None:
+            raise ValueError("Only one of source_template_path or source_template can be set.")
+        elif source_template_path is None and source_template is None:
+            raise ValueError("One of source_template_path or source_template must be set.")
+
+        self.destination_path = Path(destination_path)
+        self._source_template_path = source_template_path
+        self._source_template = source_template
+        self._context = context
+        self.user = user
+        self.group = group
+        self.permissions = permissions
+
+    @property
+    def context(self):
+        """Returns the context_function input, rendered to a dict."""
+        if callable(self._context):
+            return self._context()
+        return self._context or {}
```

Comment:
I've added this to the docstring - does that make it clearer?  Its because when using this with `Component`s in our `CharmReconciler`, we set these up during `__init__` but some of the data these need doesn't exist yet (for example, if component1 generates the context needed for component2).  So these all accept callables so the inputs can be lazily generated

Reply:
Great explanation, the docstring explains it very well too. Thanks!